### DISCUSSION TITLE
Track Usage Data Internally

### DISCUSSION THREAD

**Discussion by benhaynes** [directus/directus#8434] [Feature Requests]
Some metrics I would like to track or have access to within Directus:

* API Requests (totals, but also over time)
* Total Database Space (eg: MB)
* Auth Requests (totals, but also over time)
* Storage Requests (totals, but also over time)
* Total Storage Space (eg: MB)

One way to store this is based on a frequency (eg: every 15m), SUMing values within the window. It would also be nice if the instance could track hard/soft maximums in the ENV file.

**Comment by massyah** [comment]
Hi @benhaynes, I was looking exactly for this. 
Instead of adding storage and aggregation logic into Directus, would you consider an approach where Directus exposes a /metrics endpoints in OpenMetrics format ? 
That would instantly make Directus compatible with various telemetry solutions (Datadog, prometheus, etc.)
I'd be happy to contribute if you think this is useful.

**Reply by rijkvanzanten** [reply]
I think OpenMetrics could be a really good `?export` extension type! I'd love to have a way to return any collection in the OpenMetrics format 🙂  (cc @Oreilles who had the same idea for GeoJSON)

**Comment by Jason-CKY** [comment]
Hi, is there any updates on this? Was looking for a way to monitor Directus API metrics via prometheus. I think it would be great to expose a `/metrics` endpoint in OpenMetrics format. This could allow teams to add directus monitoring into their existing monitoring solutions setup!

**Reply by rijkvanzanten** [reply]
What would `/metrics` return for use with prometheus?

**Reply by azrikahar** [reply]
@Jason-CKY There's no such built-in endpoint at this point in time, but since we can tap into the Express.js instance via hooks, it should be very much possible to use a related package to make this endpoint available!

Did a quick search and I've used [express-prom-bundle](https://www.npmjs.com/package/express-prom-bundle) for this example.

By combining the "Sample Usage" shared in said package and the [Init hook](https://docs.directus.io/extensions/hooks/#init-events) in Directus, we can then create a simple extension that'll look like this:

```js
const promBundle = require("express-prom-bundle");

module.exports = ({ init }) => {
	init("routes.before", ({ app }) => {
		app.use(promBundle({ includeMethod: true, includePath: true }));
	});
};
```

Note: I included `includeMethod` and `includePath` so that we can see the method and path for demo purpose, but do consult their docs to find out what options are available.

The reason why we're using `routes.before` init hook is because the routes should be registered _after_ this middleware, as described by their docs:

> The order in which the routes are registered is important, since **only the routes registered after the express-prom-bundle will be measured**

After installing `prom-client` and `express-prom-bundle` (once again do refer to their docs) and adding this extension to `/extensions/hooks/metrics/index.js`, we will see that there will now be a `/metrics` endpoint with all the relevant metrics for monitoring:

https://user-images.githubusercontent.com/42867097/180642502-365106c7-7a07-4494-a0df-2d731cc4c3ed.mp4

**Reply by Jason-CKY** [reply]
Thanks for the information! That would be very helpful in exposing common connection metrics that prometheus would be able to crawl from.

**Comment by superholz** [comment]
Hi, I'm trying to connect directus in running in kubernetes to datadog. Should it be possible already? This posts does not sound optimistic. But it should be possible to start the tracer library somehow. I'm exploring to do that in an extension or in the dockerfile based on datadog [documentation](https://docs.datadoghq.com/tracing/trace_collection/dd_libraries/nodejs/?tab=containers). So far no success, if someone has a hint, that would be great.

**Reply by Jason-CKY** [reply]
You can write the extension @azrikahar wrote, and bake the folder into a custom directus docker image to run in kubernetes. I wrote a minimal example [here](https://github.com/Jason-CKY/directus-deployment). The example i wrote involves exposing a /metrics endpoint that is compatible for prometheus to crawl. In your case you could add the tracer extension instead and it should work the same way.

**Reply by superholz** [reply]
Hi, thank you for the hint. I tried something similar. Like creating an extension and trying to init the tracer in the executed function like this
```
import { defineHook } from '@directus/extensions-sdk';
import tracer from 'dd-trace';

export default defineHook(({ init }, { logger }) => {
	init('app.before', () => {
		logger.info("Init datadog tracer...")
		tracer.init();
	});
});
```

or this

```
import tracer from 'dd-trace';
tracer.init(); 
export default tracer;
```

Both versions show no error when building the extension. But when starting directus with docker I get the errors in the console:

`  Couldn't register hook "datadog"`
`Class extends value undefined is not a constructor or null`

Using similar syntax as @azrikahar gave me troubles in building the extension. I think that is due to the fact that the normal `export default defineHook` does not allow to tab into the express `app`. That's probably only possible with the `module.export` not using the `defineHook` function? There I got stuck to get the extension build. I could not get to work to use a .js file instead of a .ts file. Any hint here would be nice.

**Reply by superholz** [reply]
Something like this gives no problems when building the extension 

```
import { defineHook } from '@directus/extensions-sdk';
import tracer from 'dd-trace';

export default defineHook(({ init }, { logger }) => {
	init('routes.before', ({ app }) => {
		logger.info("Init datadog tracer...")
		app.use(tracer.init());
	});
});

```

but when starting directus I get the same:
`Couldn't register hook "datadog"`
`Class extends value undefined is not a constructor or null`

**Reply by Jason-CKY** [reply]
I was able to register this hook after scaffolding with `npm init directus-extension` tool, and the `index.js` containing:

```
const tracer = require('dd-trace');

export default ({ init }) => {
	init("app.before", ({ app }) => {
		console.log("Initialize datadog tracer...")
		tracer.init();
		console.log("Initialized datadog tracer.")
	});
};
```

![image](https://user-images.githubusercontent.com/27609953/181904324-dc3404b1-46c1-435a-845b-ab45d1f45efb.png)

**Reply by superholz** [reply]
It is very promising that you found a way to load it @Jason-CKY !

I tried your code and made sure I did the scaffolding of the extension with language option `javascript`.

However it still is not working. I get the following log output. First it seems the extension got registered properly but it cannot find the library.
 
<img width="774" alt="image" src="https://user-images.githubusercontent.com/71103630/182002620-d79d4631-76f9-4f07-9f58-f8415b25760a.png">

My package.json of the extension looks like this:

```
{
	"name": "directus-extension-datadog",
	"version": "1.0.0",
	"keywords": [
		"directus",
		"directus-extension",
		"directus-custom-hook"
	],
	"directus:extension": {
		"type": "hook",
		"path": "../../dist/hooks/datadog/index.js",
		"source": "src/index.js",
		"host": "^9.14.4"
	},
	"scripts": {
		"build": "directus-extension build",
		"dev": "directus-extension build -w --no-minify"
	},
	"devDependencies": {
		"@directus/extensions-sdk": "9.14.4"
	},
	"dependencies": {
		"dd-trace": "^2.12.1"
	}
}

```

I tried with the auto dev auto load extension on or off.

**Reply by superholz** [reply]
@Jason-CKY did register your tracer as a hook - so also located in 'extensions/hooks' folder or other location?

**Reply by superholz** [reply]
@Jason-CKY just wondering what that line in your logs refer to - has it to do with this?
`Applying Rename Hook Trigger to Event.`

**Reply by Jason-CKY** [reply]
Yes. Run `npm run build`, then move the resulting `index.js` file from the `./dist` folder to the `./datadog` folder generated. Then register it by mapping `./datadog:/directus/extensions/hooks/datadog` volumes

**Reply by superholz** [reply]
Hmm maybe I' missing something. My extension I placed in `app/extensions/hooks/datadog`. In the package.json of the extension I have `"path": "../../dist/hooks/datadog/index.js",`. So the resulting index.js file ends up in `app/extensions/dist/hooks/datadog/index.js`. Further, in the docker-compose I have the mapping `./app/extensions/dist:/directus/extensions`. I would think that this should do equally good as you mentioned in the last post. Or do you see something that should be different?

**Reply by superholz** [reply]
I have the suspicion that the way the build is configured in the Dockerfile may infringe somehow with the approach building outside and map the dist directory. That's the content of the Dockerfile is 
```
FROM directus/directus:9.13.0

USER root

# Install some extra tools for easier developing
RUN apk add --no-cache git curl zsh
RUN sh -c "$(curl -fsSL https://raw.github.com/ohmyzsh/ohmyzsh/master/tools/install.sh)"

# Build extensions
COPY ./app/extensions ./build/extensions
RUN cd ./build/extensions && ./build_extensions.sh
RUN cp -R ./build/extensions/dist/. ./extensions/

# Include migrations
COPY ./app/migrations/migrations ./extensions/migrations

CMD npx directus bootstrap && \
    npx directus start
```

**Reply by superholz** [reply]
So far I got a minimal example with a new repo working and try to work backwards for my actual application.

**Reply by superholz** [reply]
Current final problem/solution was that for the approach for my repo the extension node_modules where not available to the extension after building them. I changed folder structure to have them avaliable. Still not sure why this was no problem for the extensions that were added earlier.

**Reply by marcelolima** [reply]
@superholz have you had success doing it? I'm able to call `tracer.init()` properly in the init hook, but it doesn't seem datadog can tap into expressjs as expressjs is imported before the event app.before is emitted.

**Reply by sebastian-holzapfel** [reply]
I remember having it working with the tweeks mentioned above. It's already some time ago and I no longer have access to my solution.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by NilsBaumgartner1994** [reply]
Any updates on metrics support?

**Comment by AmitBar2001** [comment]
any updates? we also need this

**Reply by NilsBaumgartner1994** [reply]
Directus now supports prometheus

**Reply by marcelolima** [reply]
Hey @NilsBaumgartner1994 what about datadog? The workaround with init hook using `app.before` makes it possible to call `tracer.init()` but this happens after expressjs is [imported](https://github.com/directus/directus/blob/fd5a22299f4a8659e66b28eef42b6aab7e907be7/api/src/app.ts#L178) in app.ts, hence dd-trace doesn't seem able to tap into it. Would there by any alternative?
